portfolio/render.go

```diff
@@ -218,3 +225,156 @@ func RenderAggregateCapitalGains(
+	day    date.Date              // the date at which the yearly max value was seen
+	total  decimal_opt.DecimalOpt // the total ACB of all securities on that date
+	values []string               // the securities values for the related row from "total costs"
+}
+
+// RenderTotalCosts generates two RenderTable instances in a CostsTable to determine the maximum
+// total cost of all included securities at any point during the year, and then to find the maximum
+// value of each year.
+//
+// We do this by looking at any date with one or more TxDelta settlements. We take the maximum for the
+// day (in the case of multiple settlements on a given day for the same security), and then add that to
+// the current ACB for all the other tracked securities on that day. If there isn't a TxDelta settlement
+// for any of the securities on a given day, we use the last value seen. For this reason we must process
+// all the TxDeltas in date order.
+//
+// Examples:
+//
+//	Total:
+//
+//		     DATE    |  TOTAL  |  SECA   |  XXXX
+//		-------------+---------+---------+---------
+//		  2001-01-13 | $100.00 | $100.00 | $0.00
+//		-------------+---------+---------+---------
+//		  2001-02-14 | $190.00 | $100.00 | $90.00
+//		-------------+---------+---------+---------
+//		  2001-03-15 | $90.00  | $0.00   | $90.00
+//		-------------+---------+---------+---------
+//		  2001-04-16 | $80.00  | $0.00   | $80.00
+//		-------------+---------+---------+---------
+//		  2001-05-17 | $270.00 | $200.00 | $70.00
+//		-------------+---------+---------+---------
+//		  2003-01-01 | $70.00  | $0.00   | $70.00
+//		-------------+---------+---------+---------
+//
+//
+//	Yearly:
+//
+//		  YEAR |    DATE    |  TOTAL  |  SECA   |  XXXX
+//		-------+------------+---------+---------+---------
+//		  2001 | 2001-05-17 | $270.00 | $200.00 | $70.00
+//		-------+------------+---------+---------+---------
+//		  2003 | 2003-01-01 | $70.00  | $0.00   | $70.00
+//		-------+------------+---------+---------+---------
+func RenderTotalCosts(allDeltas []*TxDelta, renderFullDollarValues bool) *CostsTables {
+	dateCosts := map[date.Date]*util.DefaultMap[string, decimal_opt.DecimalOpt]{}
+
+	// For the rendered tables, we need all of the security tickers / names.
+	securitySet := map[string]struct{}{}
+
+	// Keep track of the maximum cost for each security on any date where there's a TxDelta.
+	// For example, SECA on 2000-01-01 has ACB 12, ACB 150, and ACB 0, so after the loop below,
+	// we'll have a dateCosts[2001-01-01][SECA] = 150
+	for _, d := range allDeltas {
+		dateFromDelta := d.Tx.SettlementDate
+		if _, ok := dateCosts[dateFromDelta]; !ok {
+			dateCosts[dateFromDelta] = util.NewDefaultMap(func(string) decimal_opt.DecimalOpt {
+				return decimal_opt.Null
+			})
+		}
+		sec := d.PostStatus.Security
+		securitySet[sec] = struct{}{}
+
+		val := dateCosts[dateFromDelta].Get(sec)
+		if val.IsNull {
+			val = decimal_opt.Zero
+		}
+		curMax := decimal_opt.Max(val, d.PostStatus.TotalAcb)
```

Comment:
So, couple things here.

1. If you have transactions from a registered account, Max is going to set this to Null, so your yearly max will be null (I think).
2. If you have multiple non-registered Affiliates, (ie your wife), the ACBs will end up conflicting, which isn't super userful.

I think it's probably _fine_ to just say this feature is limited to your default non-registered account, but we should probably filter that there.

Reply:
Good point. I think I've handled this by simply checking for `d.PostStatus.TotalAcb.IsNull` and skipping such Deltas, and printing warnings for those. Do you this (with the other caveat in the help text) is sufficient?

Reply:
If you want the help text to be accurate and also exclude non-registered non-default affiliates, you'll also need another case like:

```
} else if d.Tx.Affiliate.Default() {
   // note for non-default affiliate
```

Reply:
Yes, good catch, thanks. I think I understand Affiliates now :smile: